doc/dev/ragged_iter_domain_design_doc.md

```diff
@@ -0,0 +1,854 @@
+);
+```
+
+**Semantics:**
+- Input: Batch IterDomain (extent K) and RaggedIterDomain with K nested domains
+- Output: Regular IterDomain with extent = sum of all nested domain extents
+- This concatenates all ragged components into a single contiguous dimension
+
+**Example:**
+```cpp
+// Input: batch_id with extent 3, ragged_id with extents [127, 0, 198]
+auto merged = IterDomain::merge(batch_id, ragged_id);
+// Output: Regular IterDomain with extent 325 (= 127 + 0 + 198)
+```
+
+**Naming Note:**
+We considered using a new name such as `combine` for this operation, since it differs from the usual `merge` that only works with regular IterDomains (merging a batch IterDomain with a RaggedIterDomain is a distinct operation semantically). However, we decided to use the name `merge` to maintain consistency with the existing IterDomain transformation API, as the operation conceptually performs a merge that collapses two dimensions into one.
+
+#### Key Usage Patterns
+
+The `partition` and `merge` operations are inverse primitives for creating and flattening ragged structures:
+
+**Example - Complete workflow:**
+```cpp
+// Start: Flattened tokens [325 tokens, 512 hidden_dim]
+auto token_dim = IrBuilder::create<IterDomain>(0, 325);
+
+// Forward: Partition into 3 experts with extents [127, 0, 198]
+// offsets = [0, 127, 127, 325]
+auto [expert_dim, tokens_per_expert] = IterDomain::partition(token_dim, offsets);
+// expert_dim: IterDomain with extent 3
+// tokens_per_expert: RaggedIterDomain with nested extents [127, 0, 198]
+
+// ... process with ragged structure ...
+
+// Reverse: Merge back to flattened form
+auto merged_dim = IterDomain::merge(expert_dim, tokens_per_expert);
+// merged_dim: Regular IterDomain with extent 325
+```
+
+**Usage patterns:**
+
+1. **Representing inherent data structure**: Use `asNested` (which internally uses `partition`) to represent ragged dimensions that are intrinsic properties of the user data (e.g., variable tokens per expert)
+
+2. **Scheduling decisions**: Use `partition` directly for multi-GPU distribution strategies where work is split non-uniformly across devices
+
+3. **Merging back**: Use `merge` to convert ragged dimensions back to regular dimensions when needed (e.g., after expert processing, concatenate results back to merged form)
+
+The `partition` and `merge` operations enable ragged structures as shown in the examples below.
+
+#### Tensor Domain Structure
+
+A `TensorDomain` can contain a mix of regular `IterDomain` and `RaggedIterDomain` instances, representing tensors with both uniform and ragged dimensions. nvFuser may support **multiple RaggedIterDomains per TensorDomain** (nice-to-have feature, may be dropped if complicated).
+
+**Example 1: Single ragged dimension**
+```cpp
+// Batch of sequences with variable lengths
+auto batch = IrBuilder::create<IterDomain>(0, 3);
+auto ragged_seq = IrBuilder::create<RaggedIterDomain>({seq0, seq1, seq2});
```

Comment:
To save your work, I think for GroupedMmaOp and EP all we need is to asNested to "split" a dense tensor into a nested tensor. This API to create RaggedIterDomain from sequences is likely unnecessary.